feat: support for generating charmcraft YAML from Python classes

tonyandrewmeyer opened the pull request:
A new package, `ops-tools` is added. This currently only provides a script to update `charmcraft.yaml` based on config and action Python classes, but is designed to also offer other tools in the future. It is released simultaneously with the other `ops-*` packages, but is not required by any of the others, or offered as an extra.

Charmers can use the exported functions (which take a Python class object and return a dictionary suitable for serialisation to YAML), but are expected to use the provided script (that works with a `charmcraft.yaml` file and Python modules).

Classes can provide a `to_juju_schema()` method if they need to provide YAML in a different way. The script will pass the generated config/action as a base to optionally work from.

The implementation is designed to work with the `load_config` and `load_params` functionality in ops. In particular, that means working with four types of class:

* A generic Python class (no inheritance other than object).
* Standard library dataclasses (I personally believe these are the best choice for almost all cases).
* Pydantic dataclasses
* Pydantic 2.x BaseModel classes

Only reference documentation is included in this PR ([Preview](https://canonical-ubuntu-documentation-library--1975.com.readthedocs.build/ops/1975/reference/ops-tools/)). The intention is that there will be a follow-up PR that includes at least how-to documentation.

More details are available in the [spec](https://docs.google.com/document/d/1vj2pTvVRXClCrYM7tUjft3ngISSsFdBQD1B-CzOG1Vs/edit?usp=sharing). This PR replaces #1702.

tonyandrewmeyer committed:
Initial implementation.

tonyandrewmeyer committed:
Merge origin/main.

tonyandrewmeyer committed:
Tox is green, including docs.

tonyandrewmeyer committed:
Add extra tests.

tonyandrewmeyer committed:
Tox is green again.

tonyandrewmeyer committed:
Add a Raises section.

tonyandrewmeyer committed:
Expand to-do list.

tonyandrewmeyer committed:
Adjustments after initial review.

tonyandrewmeyer committed:
Provide pydantic for static checks.

tonyandrewmeyer committed:
Changes from initial review.

tonyandrewmeyer committed:
Merge from main.

tonyandrewmeyer committed:
Tweaks after merge.

tonyandrewmeyer committed:
Add TODO.

tonyandrewmeyer committed:
Remove ignore used during dev.

tonyandrewmeyer committed:
Add TODO note.

tonyandrewmeyer committed:
Do the status set right at the end, so that it only happens if charms don't catch the error themselves.

tonyandrewmeyer committed:
Tweak comment.

tonyandrewmeyer committed:
Use cleaner methods to get the set of fields.

tonyandrewmeyer committed:
Only pass the config that the class defines. This makes it easier to break the config into multiple classes.

tonyandrewmeyer committed:
Provide lazy Secrets in the config - this means you get errors later if the secret isn't available, but avoids Juju calls when loading config.

tonyandrewmeyer committed:
With lazy Secrets, we no longer get errors with missing secrets/invalid secret IDs.

tonyandrewmeyer committed:
Clean up the attr-to-type code.

tonyandrewmeyer committed:
Add a test for the 'only load some config' case.

tonyandrewmeyer committed:
Move TODO to a better location.

tonyandrewmeyer committed:
Merge origin/main.

tonyandrewmeyer committed:
Add tests for uncaught InvalidSchemaError.

tonyandrewmeyer committed:
Fix docs.

tonyandrewmeyer committed:
Adjustments from spec review.

tonyandrewmeyer committed:
Merge origin/main.

tonyandrewmeyer committed:
Provide Python 3.8 support.

tonyandrewmeyer committed:
Apply suggestions from code review

Co-authored-by: James Garner <james.garner@canonical.com>

tonyandrewmeyer committed:
Remove methods for customising attribute names, per review.

tonyandrewmeyer committed:
Sort the field names for consistent output.

tonyandrewmeyer committed:
Move private methods from the class to the module, per review.

tonyandrewmeyer committed:
Provide a more explicit type, per review.

tonyandrewmeyer committed:
Get the default value via pydantic and dataclasses if needed.

tonyandrewmeyer committed:
Improve comment.

tonyandrewmeyer committed:
Simplify the BaseModel case.

tonyandrewmeyer committed:
Use fields() per review.

tonyandrewmeyer committed:
Update ops/_private/attrdocs.py

Co-authored-by: James Garner <james.garner@canonical.com>

tonyandrewmeyer committed:
Add comment so that when 3.8 support is dropped this is found.

tonyandrewmeyer committed:
Minor tweak per review.

tonyandrewmeyer committed:
Update ops/charm.py

Co-authored-by: Ben Hoyt <benhoyt@gmail.com>

tonyandrewmeyer committed:
Remove the custom exception.

tonyandrewmeyer committed:
Reword docstring.

tonyandrewmeyer committed:
Merge origin/main

tonyandrewmeyer committed:
Tidy up after merge.

tonyandrewmeyer committed:
Small cleanup.

tonyandrewmeyer committed:
Merge origin/main.

tonyandrewmeyer committed:
Shuffle things around.

tonyandrewmeyer committed:
Merge origin/main.

tonyandrewmeyer committed:
More progress.

tonyandrewmeyer committed:
Remaining cleanup.

tonyandrewmeyer committed:
Add a script that updates charmcraft.yaml.

tonyandrewmeyer committed:
Merge origin/main.

tonyandrewmeyer committed:
Add support for releasing ops-tools.

tonyandrewmeyer committed:
Remove ignored files.

tonyandrewmeyer committed:
Use pydantic for examples.

tonyandrewmeyer committed:
Fix stripping 'action' from the class name.

tonyandrewmeyer committed:
No need to have tools installed for the integration tests at this point.

tonyandrewmeyer committed:
Make it more convenient to pass in the modules and classes.

tonyandrewmeyer committed:
Improve the readme with an example to run.

tonyandrewmeyer commented on testing/src/scenario/state.py:
```diff
@@ -1975,6 +1975,10 @@ def autoload(charm_type: type[CharmBase]) -> _CharmSpec[CharmType]:
             # try to load using legacy metadata.yaml/actions.yaml/config.yaml files
             meta, config, actions = _CharmSpec._load_metadata_legacy(charm_root)
 
+        # TODO: ideally, we would look for ConfigBase classes in the charm
+        # module and autoload from there at this point. Leaving this until the
+        # conversation about if & how the generation is done is resolved.
+
```
My feeling at the moment is that this should be left for future work, rather than baking anything in to Scenario at this time. There is a simple example of using this functionality in tests for both config and actions in the Scenario tests - I think those show that it's not too arduous to use at the moment, without anything else.

tonyandrewmeyer committed:
Tweaks from the old review.

tonyandrewmeyer committed:
Fix docs.

tonyandrewmeyer commented on tools/src/ops_tools/_attrdocs.py:
```diff
@@ -0,0 +1,103 @@
+            ):
+                docs[field.name] = field.default.description  # type: ignore
+
+    for schema_class in cls.mro():
+        if schema_class is object:
+            continue
+        try:
+            source_code = inspect.getsource(schema_class)
+        except OSError:
+            logger.debug('No source code found for %s', schema_class.__name__)
+            continue
+        try:
+            tree = ast.parse(source_code)
+        except (SyntaxError, IndentationError):
+            logger.debug('Failed to parse source code for %s', schema_class.__name__)
+            continue
+        extractor = AttributeDocstringExtractor()
+        extractor.visit(tree)
+        docs.update(extractor.attribute_docs)
```
It's exercised in at least one test, because I added it to fix a test. It is only for a single inheritance layer (plus object). If your config/action (non-Pydantic) class inherits from another class, we need to get descriptions from the parent as well, or they'll be missing from the generated YAML.

It was clobbering `model_fields` and `dataclasses.fields before the `mro` part was there, even in the original. I'm not sure what's correct here, although it seems unlikely that you would have a class with clobbering. I guess if there is a Pydantic description then it should win - should we *only* look for Pydantic descriptions if `model_fields exists? And the same for Pydantic dataclasses? I could be convinced of that.

However, you're right about the order being backwards. In the test (and probably any real use) it's additive rather than replacing attributes, so I missed that.

tonyandrewmeyer commented on tools/src/ops_tools/_update_charmcraft_yaml.py:
```diff
@@ -0,0 +1,102 @@
+    if actions:
+        charmcraft_yaml['actions'] = actions
+
+    with open(args.charmcraft_yaml, 'w') as raw:
+        yaml.safe_dump(charmcraft_yaml, raw)
```
I'd like to do that, in a `charmcraft` part. However, that's not possible right now, because of when the separate YAMl files are created. The deal was that I would get this working without any charmcraft changes and then go back to them to advocate for a nicer UX with changes.

(I could double-check this - it's my recollection of the discussions but maybe there is hackery that would make it work).

I don't like the idea of use of this tool forcing charmers to go backwards and use individual config and action files rather than the unified one.

I don't like the invasive editing pyyaml does (I'm more sad about ordering than comments), but I couldn't find any way to avoid it. I could switch to a different yaml library that does this - I guess we don't need to be too strongly influenced by ops, since it's a separate tool, although keeping the team's dependency count down as a whole is desirable.

Another option that has been discussed is instead of editing charmcraft.yaml like this, it could diff the config and actions and pretty print the result and exit non-zero if there's a difference. So it's more of a linter for now, and could be used to manually copy over differences if desired. Would you prefer that?

dimaqq commented:
> * Pydantic dataclasses
> * Pydantic 2.x BaseModel classes

IIRC these two actually differ in minutae when class is instantiated.

It's something along the lines of:
- pydantic dataclass can be instantiated with wrong types, doesn't automatically raise ValueError (?)
- pydantic BaseModel coerces types, for example either str --> int or int --> str, sorry I don't recall which way, I think each field applies own type as "constructor" with raw value as the parameter

A decent set of tests is needed here, even if it feels like we're testing Pydantic and now our code.

tonyandrewmeyer commented:
> IIRC these two actually differ in minutae when class is instantiated.

None of the classes are actually instantiated in the YAML generation.

> A decent set of tests is needed here, even if it feels like we're testing Pydantic and now our code.

There are tests for both pydantic dataclasses and BaseModel subclasses. For BaseModel with actions, I don't think we need much more than what we have, because the bulk of the work is done by pydantic rather than this code. For the others, if you'd like to suggest specific cases to check I'm happy to add additional tests.

benhoyt commented on tools/README.md:
````diff
@@ -0,0 +1,90 @@
+For example, if your charm contains a single config class in `src/charm.py` called `Config` and three actions, which all end with 'Action' (and no other classes have that name):
+
+```bash
+update-charmcraft --config-class Config --action-class '.+Action'
````
`update-charmcraft` seems like a very generic name. Should we include "schema", like `update-charmcraft-schema`?

benhoyt commented on tools/README.md:
```diff
@@ -0,0 +1,90 @@
+# Ops Tools
```
Should this move to a proper docs page instead? I see we have a reference page for [`ops_tools`](https://canonical-ubuntu-documentation-library--1975.com.readthedocs.build/ops/1975/reference/ops-tools/#module-ops_tools), but not for the script.

I notice it's listed as `ops_tools` -- should it be namespaced as `ops.tools` like our other packages?

Also, do we need "how to install it"? I'd be fine if we just recommend `uvx ops.tools.update-charmcraft` or whatever, but I think we should at least mention it.

tonyandrewmeyer replied:
Yeah, my intention is to do a follow-up PR with actual documentation, and a bunch of what's in the README probably goes into a how-to guide I think. I'm not sure if the script should be in the reference section or not, but hammering that out in the doc PR would work I think.

james-garner-canonical replied:
I was thinking about this as well. Currently, it's very different to the other packages, but rightfully so -- they're available under `ops.testing/tracing`, at testing/runtime, while these tools aren't intended to be.

Is a namespace package possible here? (Distribute as `ops-tools`, import as `ops.tools`, maybe do an extra ...).

I'm not sure it's worth shimming them in the same way we did `testing` and `tracing` (and then presumably distributing the package as `ops-tools` and making it available as an extra as `ops[tools]`).  But maybe it is? Consistency is nice, our users are familiar with this approach, and Tiexin's release script makes the version bumping and dependencies pretty painless for releases.

benhoyt replied:
> That didn't feel necessary here, particularly since I think people will just be running the script most of the time rather than actually doing an import.

Yeah, that's fine. I'm happy to leave that as is.

benhoyt replied:
> My sense was that we wanted this to feel a little less baked in - these are tools to use with ops, not something that's part of ops itself.

Yeah, I agree with that, and on reflection would prefer to avoid `ops.tools` for that reason.

benhoyt reviewed (approved):
Couple of relatively minor comments, but approving optimistically. I haven't reviewed the code line-by-line (I did that in the previous incantation, and I'm also less concerned as it's an offline tool).

tonyandrewmeyer committed:
Fix the order of moving through the classes, add an explicit test for this.

tonyandrewmeyer committed:
Merge origin/main

tonyandrewmeyer committed:
Make the entry point name clearer.

tonyandrewmeyer committed:
Allow merging rather than replacing.

tonyandrewmeyer committed:
Add a diff method.

tonyandrewmeyer committed:
WiP doctests.

tonyandrewmeyer committed:
Edit the charmcraft.yaml in a way that preserves comments and ordering.

tonyandrewmeyer committed:
Update for newer pyright.

tonyandrewmeyer committed:
Merge origin/main.

tonyandrewmeyer committed:
Fix order of output.

benhoyt commented on tools/src/ops_tools/_update_charmcraft_yaml.py:
```diff
@@ -0,0 +1,191 @@
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    parser.add_argument(
+        '--charmcraft-yaml',
```
This is the path to the main artifact we're changing. What about calling this arg `--path` or similar? No need to be overly verbose.

benhoyt commented on tools/src/ops_tools/_update_charmcraft_yaml.py:
```diff
@@ -0,0 +1,191 @@
+        default='charmcraft.yaml',
+    )
+    parser.add_argument(
+        '--config-class',
```
I realise this is more accurate, but it's a bit long. Would `--config` and `--action` be enough?

james-garner-canonical commented on tools/src/ops_tools/_generate_yaml.py:
```diff
@@ -0,0 +1,434 @@
+}
+
+# We currently only handle the basic types that we expect to see in real charms.
+# lists and tuples (arrays) are handled without using this mapping.
```
We do handle dicts too, right?
```suggestion
# JSON arrays and objects (lists/tuples and dicts) are handled without using this mapping.
```

james-garner-canonical commented on tools/src/ops_tools/_generate_yaml.py:
```diff
@@ -0,0 +1,434 @@
+}
+
+
+def attr_to_default(cls: type[object], name: str) -> Any:
```
```suggestion
def attr_to_default(cls: type[object], name: str) -> object:
```
We explicitly check the types after calling this, so we should be fine to use `object` instead of `Any` for better type safety, right?

james-garner-canonical commented on tools/src/ops_tools/_generate_yaml.py:
```diff
@@ -0,0 +1,434 @@
+        return getattr(cls, name, None)
+    default = None
+    for field in dataclasses.fields(cls):
+        if field.name != name:
+            continue
+        break
```
```suggestion
        if field.name == name:
            break
```

james-garner-canonical commented on tools/src/ops_tools/_generate_yaml.py:
```diff
@@ -0,0 +1,434 @@
+    """Get the default value for the attribute."""
+    if not dataclasses.is_dataclass(cls):
+        return getattr(cls, name, None)
+    default = None
+    for field in dataclasses.fields(cls):
+        if field.name != name:
+            continue
+        break
+    else:
+        return default
```
This might be easier to follow if we do something like this:
```suggestion
    fields = {field.name: field for field in dataclasses.fields(cls)}
    if name not in fields:
        return None
    field = fields[name]
```

tonyandrewmeyer replied:
I was trying to avoid building up a dict just to get a single item out of it. With the fix from above, I think it's better than adjusting it this way.

(It's really annoying that there's no dataclasses.get_field, or that dataclasses.fields isn't a dict).

dimaqq replied:
My 2c: if dataclasses doesn't provide a dict, a helper method that builds a dict is the most natural way to go.

james-garner-canonical commented on tools/src/ops_tools/_generate_yaml.py:
```diff
@@ -0,0 +1,434 @@
+        return field_default  # type: ignore
+    if field_default_factory is not dataclasses.MISSING:
+        return field_default_factory()  # type: ignore
+    return default
```
We don't actually set this anywhere other than initially setting it to `None`, right? I suggest not setting it at all (in previous suggestion), and I think it's easier to follow the function without it (you don't have to wonder if it's being modified at any point).
```suggestion
    return None
```

james-garner-canonical commented on tools/src/ops_tools/_generate_yaml.py:
```diff
@@ -0,0 +1,434 @@
+    try:
+        raw_hint = get_type_hints(cls)[name]
+    except KeyError:
+        pass
+    else:
```
```suggestion
        return 'string'
```
And dedent the rest of the function accordingly. Though here rather than having multiple occurences of the possible-to-type string value `'string'`, I would prefer using a constant, e.g. `default: Final[str] = yaml_type[str]`.

james-garner-canonical commented on tools/src/ops_tools/_generate_yaml.py:
```diff
@@ -0,0 +1,434 @@
+        origin = get_origin(raw_hint)
+        if origin in (list, tuple):
+            return yaml_types[origin]
+        elif origin:
```
```suggestion
        if origin:
```

james-garner-canonical commented on tools/src/ops_tools/_generate_yaml.py:
```diff
@@ -0,0 +1,434 @@
+    else:
+        # Collapse Optional[] and Union[] and so on to the simpler form.
+        origin = get_origin(raw_hint)
+        if origin in (list, tuple):
+            return yaml_types[origin]
```
We call this function with `yaml_types=JSON_TYPES` and `yaml_types=JUJU_TYPES`, neither of which contain `list` or `tuple`. I'm assuming that `list` and `tuple` are handled correctly elsewhere. Maybe the right thing to do here is raise an explicit error (instead of the `KeyError` we'd get in practice).

tonyandrewmeyer replied:
Nice catch. This is leftover from earlier code that was handling the list/tuple cases. That's done more cleanly (although with a bit of duplication, annoyingly) for actions now before this is called, so it'll never be triggered in that case.

For config, we don't have a way to say "this is an array" in the Juju form, so it should fall back to string like elsewhere.

james-garner-canonical commented on tools/src/ops_tools/_generate_yaml.py:
```diff
@@ -0,0 +1,434 @@
+        origin = get_origin(raw_hint)
+        if origin in (list, tuple):
+            return yaml_types[origin]
+        elif origin:
+            hints = {arg for arg in get_args(raw_hint) if arg in yaml_types}
+        else:
+            hints = {raw_hint}
+        # If there are multiple types -- for example, the type annotation is
+        # `int | str` -- then we can't determine the type, and we fall back to
+        # "string", even if `str` is not in the type hint, because our
+        # "we can't determine the type" choice is always "string".
+        if len(hints) > 1:
+            return 'string'
+        elif hints:
+            try:
+                return yaml_types[hints.pop()]
+            except KeyError:
+                pass
```
I think this is a simpler way of writing the same logic -- with one difference: we now *always* return `'string'` if there are multiple hints, instead of making an exception for the case where exactly one of them is in `yaml_types`. To bring that logic back, filter the result of `get_args` based on `yaml_types` -- but I think the semantics are clearer this way.

```suggestion
        hints = set(get_args(raw_hint)) if origin else {raw_hint}
        # If there are multiple types -- for example, the type annotation is
        # `int | str` -- then we can't determine the type.
        # Likewise if there are no hints somehow (generic used without args?).
        # If we can't figure it out, then use "string", which should be the most
        # compatible, and most likely to be used for arbitrary types. Charms can
        # override `to_juju_schema` to adjust this if required.
        return yaml_types.get(hints.pop(), 'string') if len(hints) == 1 else 'string'
```

james-garner-canonical commented on tools/src/ops_tools/_generate_yaml.py:
```diff
@@ -0,0 +1,434 @@
+    # The many type: ignores are required because we don't want to import
+    # pydantic in this code.
+    options: dict[str, OptionDict] = {}
+    for name, field in cls.model_fields.items():  # type: ignore
```
This is as far as I've gotten in my review so far. I'm wondering here if we can reuse any of the logic already implemented in ops for extracting fields from classes, e.g. [ops.charm._juju_fields](https://github.com/canonical/operator/blob/01740d4947512afc78bcc150a96f9998e3ab24e8/ops/charm.py#L1604).

james-garner-canonical reviewed (commented):
Partial review for now.

tonyandrewmeyer committed:
Update tools/README.md

Co-authored-by: James Garner <james.garner@canonical.com>

tonyandrewmeyer committed:
Apply suggestions from code review

Co-authored-by: James Garner <james.garner@canonical.com>

tonyandrewmeyer committed:
Tweak args, per review.

tonyandrewmeyer committed:
Merge origin/main.

tonyandrewmeyer committed:
Apply suggestions from code review

Co-authored-by: James Garner <james.garner@canonical.com>

tonyandrewmeyer commented on tools/pyproject.toml:
```diff
@@ -0,0 +1,57 @@
+]
+
+[project.scripts]
+update-charmcraft-schema = "ops_tools._update_charmcraft_yaml:main"
```
I'm not that keen on having things as sub-commands under a single `ops-tools` command. For one thing, I don't like the duplication:

```
uvx --from=ops-tools ops-tools generate-metadata
```

But also I feel like `ops-tools` is meant to be a collection of tools, not a single tool with multiple subcommands.

However, I haven't finished going through the CLI spec in detail, so maybe it has recommendations here that should be followed.

I would like to resolve the comment around how this should work (just a diff/lint, merging, replacing) before locking down what the command looks like and is named.

james-garner-canonical replied:
Re: duplication, you shouldn't need any if the command name is the same as the package name, right? Just `uvx ops-tools generate-metadata`. Good call on figuring out the rest of the commandline interface first though, I agree that's more important, and I have some thoughts there too.

james-garner-canonical replied:
Leaving aside subcommands, since with uv that just means +/- the `--from` flag, I think the existing interface is nice overall. The main parts I find confusing are what will actually happen with the content we generate.

I'd suggest using mutually exclusive groups for the output options, and defaulting to a non-destructive output method like writing the generated content to stdout. How about:
- `--update` (current behaviour, write to existing file)
- `--diff` (current option)
- `--output=FILE` (default behaviour, with `-` for stdout)
I'd also be happy with a `--stdout` flag instead of `--output=` if that's simpler.

The `--merge` flag option had me a little confused too, I thought it was the option that turned on the `--update`  behaviour, but that's (currently) unconditional. Instead, it controls whether to keep actions/options that are in the `charmcraft.yaml` file but not present in our generated output, right? In that case, how about `--delete` / `--no-delete` flags for this (borrowing `rsync` semantics)?

Regarding the name, how about `update-charmcraft-yaml` instead? The term charmcraft schema seems a little off somehow. All together that gives us:
```
uvx {--from} ops-tools update-charmcraft-yaml
    --path=...  # This might be nice as an optional positional argument instead?
    --config=... ...
    --action=... ...
    [--delete / --no-delete (default)]
    [--update / --diff / --output='-' (default)]
```

james-garner-canonical replied:
It also seems like it would be nice to try this out without consuming a `charmcraft.yaml` file at all. What do you think about additional entry points (scripts or subcommands) with much simpler interfaces that just write the actions/config options to stdout? 
```
uvx {--from} ops-tools generate-actions CLASS...
uvx {--from} ops-tools generate-config CLASS...
```
I'd suggest writing in the `actions.yaml` and `config.yaml` format, but there could always be a flag added to control that.

james-garner-canonical commented on tools/src/ops_tools/_update_charmcraft_yaml.py:
```diff
@@ -0,0 +1,191 @@
+    parser.add_argument(
+        '--action',
+        action='append',
+        help='Python class with action classes (can be specified multiple times). '
```
```suggestion
        help='Python class with optional module path (can be specified multiple times). '
```

james-garner-canonical reviewed (changes requested):
I've left some major and minor comments.

Major: I think the CLI interface is pretty nice, but what actually happens with the output could be made more clear. I'd like to see non-destructive default behaviour by default, and I'd also like to be able to dump a Python class to a Juju action/config definition without worrying about `charmcraft.yaml` at all.

More minor: while reading through the code, I've left a number of suggestions that I think might make some of the logic easier to follow. However ...

Big picture questions:
1. How much simpler would interacting with Pydantic model classes be if `ops-tools` could just import Pydantic?
    - Additionally, if we're importing a module that defines a Pydantic class, won't we have to have Pydantic installed anyway?
    - We could either add Pydantic to our own deps
    - Or if we're relying on the user doing `--with=pydantic` or something, then we could move our Pydantic handling logic to a module that imports `pydantic` and only import that module as needed. This would have the benefit of letting us fudge around the Pydantic version, but seems more complicated.
2. I may be wrong about this, but my impression is that we're doing similar class traversal logic in a few different places for different purposes. This is some of the trickier code to follow imo. Would it be worthwhile to centralise this logic and parse the classes we're consuming into an intermediate data structure that has everything we want represented nicely?
    - I also had kind of assumed that the core class parsing logic would be in `ops` (internals), and we'd use it in `ops-tools`, since we already parse classes for loading actions/config/relation data in `ops`. That wouldn't work if we went down the route of using Pydantic here, but perhaps something to keep in mind.

tonyandrewmeyer committed:
Apply suggestions from code review

Co-authored-by: James Garner <james.garner@canonical.com>

tonyandrewmeyer committed:
Remove the --merge and --diff options.

tonyandrewmeyer committed:
Remove unnecessary comment.

tonyandrewmeyer committed:
Include tools in coverage.

tonyandrewmeyer committed:
Tweak comment, per review.

tonyandrewmeyer committed:
Minor readability improvement, per review.

tonyandrewmeyer commented on tools/src/ops_tools/_generate_yaml.py:
```diff
@@ -0,0 +1,437 @@
+        if hasattr(field.default, 'default_factory')
+        else field.default_factory
+    )
+    # A hack to avoid importing Pydantic here.
```
I'm pretty strongly opposed to pulling Pydantic in, even for ops-tools. It would be a bit easier, but not a lot, I feel.

james-garner-canonical replied:
May I ask why -- is it the extra runtime overhead for users?

tonyandrewmeyer committed:
Minor readability improvement, per review.

tonyandrewmeyer committed:
Minor refactor, per review.

tonyandrewmeyer committed:
Minor refactor, per review.

tonyandrewmeyer committed:
Handle lists and tuples in config.

dimaqq commented on tools/src/ops_tools/_update_charmcraft_yaml.py:
```diff
@@ -0,0 +1,191 @@
+import sys
+from typing import Any, Generator
+
+import yaml
```
IMHO you really want `ruamel.yaml` here.

I'm guessing what you do is trying to figure out the boundaries of the existing section and then replacing it wholesale relying on the fact that it's YAML and not e.g. JSON, and that it's formatted with indentation.

In that case, how about including explicit delimiters, e.g.?

```yaml
storage:
  data:
    type: filesystem
    location: /tempo-data

# ==== BEGIN GENERATED ACTIONS ====

actions:
  list-receivers:
    description: |
      Returns a list of all enabled receiver endpoints.

# ==== END GENERATED ACTIONS ====

peers:
  tempo-peers:
    interface: tempo_peers
    description: |
      peer relation for internal coordination
```

WDYT?

dimaqq commented on tools/src/ops_tools/_update_charmcraft_yaml.py:
```diff
@@ -0,0 +1,191 @@
+            exit_code += 2
+        sys.exit(exit_code)
+
+    if args.merge and 'config' in charmcraft_yaml:
```
**question:** Is this for 12-factor charms?

Personally I would like to avoid merge like a plague of rats gorged on plagued infested locusts who ate all the plagued corn...

12F could be solved by a stub / base class for options.

dimaqq commented on tools/src/ops_tools/_update_charmcraft_yaml.py:
```diff
@@ -0,0 +1,191 @@
+        config = charmcraft_yaml['config']['options'].update(config['options'])
+    raw_yaml = _insert_into_charmcraft_yaml(raw_yaml, 'config', {'config': config})
+    if actions:
+        if args.merge and 'actions' in charmcraft_yaml:
```
Are there unit tests and functional / test vector tests for the merge flag?

dimaqq commented on tools/tests_tools/test_generate_yaml_config.py:
```diff
@@ -0,0 +1,241 @@
+
+@pytest.mark.parametrize('config_class', _test_config_classes)
+def test_config_yaml_schema(config_class: type[object]):
+    generated_schema = ops_tools.config_to_juju_schema(config_class)
+    expected_schema = {
+        'options': {
+            'basic-bool': {
+                'type': 'boolean',
+            },
```
This is great, but omits YAML formatting.

There needs to be a test for verbatim YAML output.
I'm thinking a stability test perhaps (we may change the code, but the output keeps same "canonical" indent).

dimaqq commented on tools/README.md:
````diff
@@ -0,0 +1,90 @@
+would generate this YAML:
+
+```yaml
+actions:
+    run-backup:
+        description: Backup the database.
+        params:
+            filename:
+                type: string
+                description: The name of the backup file.
+            compression:
+                type: string
+                description: The type of compression to use.
+                default: gzip
+                enum: [gzip, bzip2]
+        required: [filename]
+        additionalProperties: false
````
I'd like to see a test that looks exactly like this.

dimaqq reviewed (commented):
First round of reviews.

tonyandrewmeyer committed:
Align the behaviour of juju_names exactly with Relation.save().

dimaqq reviewed (commented):
Partial review

tonyandrewmeyer committed:
Merge origin/main.

tonyandrewmeyer committed:
Merge upstream/main.

tonyandrewmeyer committed:
Drop unused setuptools_scm build dep in ops-tools.

setuptools_scm was pulled in but never used (no git-tag based
versioning) — and on modern setuptools it fails to import because
it still references the removed pkg_resources.

tonyandrewmeyer committed:
Fix ops.Secret | None detection and Pydantic enum $ref handling.

Two post-merge regressions in ops-tools:

1. _attr_to_yaml_type was treating ops.Secret | None as a multi-type union
   (keeping NoneType in the hint set) and falling back to 'string'. Strip
   NoneType so Optional[T] is treated as T, restoring 'type: secret' for
   optional secret fields.

2. Pydantic 2.12 no longer inlines Enum schemas; it emits $defs + $ref.
   action_to_juju_schema now calls model_json_schema() (schema() is
   deprecated) and inlines referenced schemas so the Juju YAML retains
   type / enum instead of a dangling $ref.

tonyandrewmeyer committed:
Address remaining review nits.

* JSON_TYPES comment wording (james-garner-canonical): tighten the
  description of what's handled outside the mapping.
* test_main charm comment (dimaqq): reword to match the style of the
  surrounding comment.
* _attrdocs unit tests (dimaqq): add two edge cases -- a class
  docstring must not attach to the first attribute, and a method
  between attributes must not break docstring collection.

tonyandrewmeyer committed:
Delegate juju_names to ops.charm._juju_fields.

Share the dataclass / Pydantic class traversal with ops.charm rather
than reimplementing it here. ops-tools stays permissive by catching
the ValueError that _juju_fields raises on plain classes and falling
back to get_type_hints.

tonyandrewmeyer committed:
Add generate-juju-config and generate-juju-actions entry points.

update-charmcraft-schema requires a charmcraft.yaml to read and write
back, which makes it awkward to try the generator interactively, pipe
the output somewhere else, or drive it from a charmcraft override-build
part (where charmcraft.yaml may not declare config/actions at all).

Add two standalone tools that take one or more Python class specifiers
and write the matching Juju YAML (config.yaml / actions.yaml format)
to stdout. They share get_class_from_module with the existing tool
(moved into _generate_juju_yaml.py) but do not read charmcraft.yaml.

tonyandrewmeyer committed:
The cycle field isn't needed any more.

tonyandrewmeyer commented:
Opening this fresh as smaller PRs with some small spec'ing.
